core/blockchain.go

```diff
@@ -1298,13 +1300,14 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 	bc.wg.Add(1)
 	defer bc.wg.Done()
 
+	if hash := bc.GetCanonicalHash(block.NumberU64()); (hash != common.Hash{} && hash != block.Hash()) {
```

Comment:
shouldn't insertChain have a similar check? Thinking here is that we want to avoid processing the block if it's already non cannonical.

Reply:
The insertChain, basically skip all the blocks that are already in the db, and then, starts to apply a `bc.writeKnownBlock(block)` to every block, that knows, but the head is pointing to a smaller block (this could be due to a rollback). This `writeKnownBlock` checks if the parent is a canon one.
As this parts of the basis that the old rolled back block was written, it means that it already passed for the validation that at the same level the block is the only one available (the new canon), that's why we are not checking it. It's basically a head update in the case of sending the same chain after a rollback, and at that level it can't exist a canon because the last canon is the parent that you are checking.
After that, the chainInsert, with the unknown blocks, calls the `WriteBlockWithState`
which checks that for the same level, no canon already exists.

So, TL;DR; we are checking that, because we are using the same function inside insertChain

Reply:
Seems to me that this check is already being done on `InsertChain` and thus this check should be moved to `WriteBlockWithState()` since that's the only other caller of this private method